src/client/components/pages/Courses/EnrollmentModal.tsx

```diff
@@ -0,0 +1,321 @@
+      .map(([fieldName, fieldValue]) => {
+        const errors: string[] = [];
+        const displayName = getDisplayName(fieldName);
+        // Match any alphabetical chars
+        if (new RegExp(/[A-Z]/i).test(fieldValue.toString())) {
+          errors.push(`${displayName} cannot contain alphabetical characters`);
+
+        // Match any non-alphanumeric characters(also allowing the - character)
+        } else if (new RegExp(/[^\dA-Z-]/gi).test(fieldValue.toString())) {
+          errors.push(`${displayName} cannot contain special characters`);
+
+        // Check that the number > 0
+        } else if (parseInt(fieldValue, 10) < 0) {
+          errors.push(`${displayName} cannot be negative`);
+        }
+        return errors;
+      })
+      .reduce((acc, val) => acc.concat(val), []);
+
+    if (errorMessages.length > 0) {
+      setSaving(false);
+      setValidationErrors(errorMessages);
+    } else {
+      setValidationErrors([]);
+      setSaving(true);
+      const results = await updateCourseInstance(
+        instance.id,
+        {
+          offered: instance.offered,
+          ...sanitizeEnrollmentData(enrollmentData),
+        }
+      );
+      onSave(results);
+      setSaving(false);
+    }
+  };
+
+  /**
+   * Called when the modal is closed. If there are any unsaved changes,
+   * a warning message appears, and the user must confirm discarding the unsaved
+   * changes in order to close the modal. If the user selects cancel, the user
+   * can continue making changes in the modal.
+   */
+  const onModalClose = () => {
+    if (changed) {
+      // eslint-disable-next-line no-alert
+      if (window.confirm('You have unsaved changes. Click \'OK\' to disregard changes, or \'Cancel\' to continue editing.')) {
+        setChanged(false);
+        onClose();
+      }
+    } else {
+      onClose();
+    }
+  };
+
+  return (
+    <Modal
+      ariaLabelledBy="enrollment-modal-header"
+      isVisible={isVisible}
+      closeHandler={onModalClose}
+    >
+      <ModalHeader
+        forwardRef={modalHeaderRef}
+        tabIndex={0}
+      >
+        <span id="enrollment-modal-header">
+          {`Enrollment for ${course?.catalogNumber}, ${currentSemester?.term} ${currentSemester.calendarYear}`}
+        </span>
+      </ModalHeader>
+      <ModalBody>
+        {
+          saving ? (<LoadSpinner>Saving Enrollment Data</LoadSpinner>) : (
+            enrollmentFields.map(({ key, name }) => (
+              <TextInput
+                key={key}
+                id={key}
+                name={key}
+                label={name}
+                value={enrollmentData?.[key]?.toString() || ''}
+                onChange={
+                  ({ target: { value } }: ChangeEvent<HTMLInputElement>) => {
+                    setChanged(true);
+                    setEnrollmentData((state) => ({
+                      ...state,
+                      [key]: value,
+                    }));
+                  }
+                }
```

Comment:
I did wonder about this approach, but I wasn't sure about the accessibility implications for a field where you type and nothing happens with no explanation as to why.

Reply:
Fair point. What about using the errorMessage field to provide an explanation with something more like this (which assumes that the `validationErrors` state object would be changed to type `EnrollmentData`):

```suggestion
                errorMessage={validationErrors[key]}
                onChange={
                  ({ target: { value } }: ChangeEvent<HTMLInputElement>) => {
                    const nonNumber = /\D/g;
                    if (nonNumber.test(value) {
                      setValidationErrors((state) => ({
                        ...state,
                        [key]: 'Field can only contain numbers',
                      }))
                    } else {
                      setValidationErrors((state) => ({
                        ...state,
                        [key]: '',
                      }))
                    }
                    setChanged(true);
                    setEnrollmentData((state) => ({
                      ...state,
                      [key]: value.replace(nonNumber, ''),
                    }));
                  }
                }
```

Then, attempting to enter a non-numeric value would show an error without updating the actual value, but would clear the error when a non-numeric value is entered.

Reply:
I wondered about this approach also. My only concern with _this_ would be whether it might cause issues for screen readers having text appearing and disappearing whilst the user is typing. I'll maybe check in with @Vittorio2015 about this.

Reply:
Okay, I've updated this in 2ec4e985f296eff457d16ca18609e7db4e5bffbd to validate as the user types, then in 170bf5dbdebec71b50657b508318acf33987b024I tweaked the error message to be more generic like we talked about in the standup this morning. Finally, in a91ad06a801199e267419183bd4d21a8559a63eb I updated the state fields to just indicate whether or not a field is valid (since the error message is the same regardless). This allowed me to simplify the validation logic significantly, as well as creating a memoized computed value from this to decide whether or not the form is valid.

On the subject of forms and form elements...it occurs to me that these fields aren't actually wrapped in a form and I was wondering if they should be or not. If so, that may present markup challenges in also enclosing the submit button within the same form. The course admin modal encloses the form fields(but not the submit button) within a form, while the meeting modal just doesn't use forms at all. What do you think?

Reply:
It looks like the correct way to do this is [using the `form` attribute on the `button`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-form) to reference the form ID - i.e:


```html
<form id="my-form" action="foo.php">
  <input name="title" />
</form>
<button form="my-form">
  Click Me
</button>
```

However, our `mark-one` buttons do not currently allow for this prop. 

```
Type '{ children: string; onClick: () => void; variant: VARIANT.BASE; form: string; }' is not assignable to type 'IntrinsicAttributes & ButtonProps & { children?: ReactNode; }'.
  Property 'form' does not exist on type 'IntrinsicAttributes & ButtonProps & { children?: ReactNode; }'.ts(2322)
```

Reply:
For now I've updated the inputs to be wrapped by `<form>` in c086de42c4fc27887158a4e717cacb342f13e2db

Reply:
I've opedned https://github.com/seas-computing/mark-one/issues/132 to track the above issue.